x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/privilege_form_calculator/privilege_form_calculator.ts

```diff
@@ -0,0 +1,295 @@
+   * @param privilegeIndex the index of the kibana privileges role component
+   */
+  public isIndependentSubFeaturePrivilegeGranted(
+    featureId: string,
+    privilegeId: string,
+    privilegeIndex: number
+  ) {
+    const primaryFeaturePrivilege = this.getEffectivePrimaryFeaturePrivilege(
+      featureId,
+      privilegeIndex
+    );
+    if (!primaryFeaturePrivilege) {
+      return false;
+    }
+    const selectedFeaturePrivileges = this.getSelectedFeaturePrivileges(featureId, privilegeIndex);
+
+    const feature = this.kibanaPrivileges.getSecuredFeature(featureId);
+
+    const subFeaturePrivilege = feature
+      .getSubFeaturePrivileges()
+      .find(ap => ap.id === privilegeId)!;
+
+    return Boolean(
+      primaryFeaturePrivilege.grantsPrivilege(subFeaturePrivilege) ||
+        selectedFeaturePrivileges.includes(subFeaturePrivilege.id)
+    );
+  }
+
+  /**
+   * Returns the most permissive effective privilege within the indicated mutually-exclusive sub feature privilege group.
+   *
+   * @param featureId the feature id
+   * @param subFeatureGroup the mutually-exclusive sub feature group
+   * @param privilegeIndex the index of the kibana privileges role component
+   */
+  public getSelectedMutuallyExclusiveSubFeaturePrivilege(
+    featureId: string,
+    subFeatureGroup: SubFeaturePrivilegeGroup,
+    privilegeIndex: number
+  ) {
+    const primaryFeaturePrivilege = this.getEffectivePrimaryFeaturePrivilege(
+      featureId,
+      privilegeIndex
+    );
+    if (!primaryFeaturePrivilege) {
+      return undefined;
+    }
+
+    const selectedFeaturePrivileges = this.getSelectedFeaturePrivileges(featureId, privilegeIndex);
+
+    return subFeatureGroup.privileges.find(p => {
+      return primaryFeaturePrivilege.grantsPrivilege(p) || selectedFeaturePrivileges.includes(p.id);
+    });
+  }
+
+  /**
+   * Determines if the indicated feature is capable of having its sub-feature privileges customized.
+   *
+   * @param featureId the feature id
+   * @param privilegeIndex the index of the kibana privileges role component
+   */
+  public canCustomizeSubFeaturePrivileges(featureId: string, privilegeIndex: number) {
+    const selectedFeaturePrivileges = this.getSelectedFeaturePrivileges(featureId, privilegeIndex);
+    const feature = this.kibanaPrivileges.getSecuredFeature(featureId);
+
+    return feature
+      .getPrimaryFeaturePrivileges({ includeMinimalFeaturePrivileges: true })
+      .some(apfp => selectedFeaturePrivileges.includes(apfp.id));
+  }
+
+  /**
+   * Returns an updated set of feature privileges based on the toggling of the "Customize sub-feature privileges" control.
+   *
+   * @param featureId the feature id
+   * @param privilegeIndex  the index of the kibana privileges role component
+   * @param willBeCustomizing flag indicating if this feature is about to have its sub-feature privileges customized or not
+   */
+  public updateSelectedFeaturePrivilegesForCustomization(
```

Comment:
nit: This also isn't doing any calculations...

Reply:
Similar to the argument I made earlier, I tried to remove such logic from the UI layer to make it easier to reason about. I can revisit also.

Perhaps "calculator" was a bad word choice.

Reply:
I don't mind extracting this back out into the component either

Reply:
🤷‍♂I defer to you :)